docs/units/type_hints.rst

```diff
@@ -0,0 +1,136 @@
+Unit-Aware Type Annotations
+***************************
+
+.. |Quantity| replace:: :class:`~astropy.units.Quantity`
+
+While Python is a dynamically typed language, there is also support for type hints (and even static typing) using the annotations syntax of `PEP 484 <https://www.python.org/dev/peps/pep-0484/>`_.
+For a detailed guide on type hints, function annotations, and other related syntax see the `Real Python Guide <https://realpython.com/python-type-checking/#type-aliases>`_. 
+Here we how Astropy |Quantity| can be used in type hints and annotations and also include metadata about the associated units.
+
+
+For this and all further examples we assume the following imports
+
+::
+
+   >>> import typing as T
+   >>> import astropy.units as u
+   >>> from astropy.units import Quantity
+   >>> from astropy.units._typing import Annotated
+
+
+.. _quantity_type_annotation:
+
+Quantity Type Annotation
+========================
+
+A |Quantity| can easily be used as a type annotation,
+
+::
+
+   >>> x: Quantity = 2 * u.km
+
+or as a function annotation.
+
+::
+
+   >>> def func(x: Quantity) -> Quantity:
+   ...     return x
+
+However, depending on the computer, this frequently throws an exception when compiling Sphinx docs. A safer method to annotate Quantities is by leveraging the :mod:`~typing` library to create a type variable (note the ``~``).
```

Comment:
Is this a problem that is going to disappear? I feel now one immediately gets told that really this is not easy and better not done... We should certainly write for the future and not worry about people with old python/numpy/sphinx (maybe have a `.. note::` with minimum version requirements?)

Reply:
I ran into this on various edge cases. I think it's resolved with dropping py3.6. I've instead changed the language to emphasize that ``T.TypeVar("", bound=X)`` I better at allowing subclasses so is the preferred means of annotation.

Speaking of which... a future PR might add a class method e.g. ``Quantity.strict_type()`` that is the same as ``Quantity[]``, but produces an annotated Quantity , not type-var so that quantity_input can know to convert input/output to that type and not allow a subclass.

Reply:
Oh! I actually had some thoughts about this below.

> part of the challenge is that we will want to support the numpy static type, not just at the input level, but more deeply since Quantity is a subclass and the numpy static type should apply to all subclasses.
> Part of the way we might address this is to introduce a formalism for unit-types.
u.Unit[<unit or physical type>] = Annotated[u.UnitBase, <unit or physical type>] is non-strict about the specific unit. Anything with the correct physical type is allowed. This works nicely for the decorator and astropy.modeling.

@mhvk, it might be worth generating a formal document on these and other details. when we try to take next steps. Maybe an APE? and talking to some NumPy folks.

1. How can we support NumPy typing?
2. and be valid static typing for static type checkers.
2. Where / how should types be stored?
3. Should ``Unit`` have a class getitem? If so, should it mimic Quantity like
    - u.Unit[<unit or physical type>] = Annotated[TypeVary(u.UnitBase), <unit or physical type>]
    - or not use the typeover
4. to `quantity_input`:
    - what is / is there a the difference  in meaning between a ``Annotated[Quantity, ...] and a ``Annotated[TypeVar(Quantity), ...]``
    -  ``u.Unit[<unit or physical type>] = Annotated[TypeVar(u.UnitBase), <unit or physical type>]`` is non-strict about the specific unit. Anything with the correct physical type is allowed. ``Annotated[u.UnitBase, <unit or physical type>]`` is strict.